File: plugins/plugin-elizacloud/typescript/cloud-providers/container-health.ts
/**
 * containerHealthProvider — Container health in agent state (private, on-demand).
 */

import type { IAgentRuntime, Memory, Provider, ProviderResult, State } from "@elizaos/core";
import type { CloudContainerService } from "../services/cloud-container";
import type { CloudAuthService } from "../services/cloud-auth";

export const containerHealthProvider: Provider = {
  name: "elizacloud_health",
  description: "ElizaCloud container health",
  dynamic: true,
  position: 92,
  private: true,

  async get(runtime: IAgentRuntime, _message: Memory, _state: State): Promise<ProviderResult> {
    const auth = runtime.getService("CLOUD_AUTH") as CloudAuthService | undefined;
    if (!auth?.isAuthenticated()) return { text: "" };

    const svc = runtime.getService("CLOUD_CONTAINER") as CloudContainerService | undefined;
    const running = svc?.getTrackedContainers().filter((c) => c.status === "running") ?? [];
    if (running.length === 0) return { text: "No running containers.", values: { healthyContainers: 0 } };

    const reports = running.map((c) => ({
      id: c.id, name: c.name,
      healthy: c.billing_status === "active",
      billing: c.billing_status,
    }));

    const healthy = reports.filter((r) => r.healthy).length;
    const text = [
      `Health: ${healthy}/${reports.length} healthy`,
      ...reports.map((r) => `  - ${r.name}: ${r.healthy ? "OK" : "UNHEALTHY"} (${r.billing})`),
    ].join("\n");

    return { text, values: { healthyContainers: healthy, unhealthyContainers: reports.length - healthy }, data: { reports } };
  },
};

File: plugins/plugin-elizacloud/python/elizaos_plugin_elizacloud/cloud_providers/container_health.py
"""
containerHealthProvider — Container health in agent state (private, on-demand).
"""

from __future__ import annotations

from typing import TypedDict

from elizaos_plugin_elizacloud.services.cloud_auth_service import CloudAuthService
from elizaos_plugin_elizacloud.services.cloud_container_service import CloudContainerService


class ProviderResult(TypedDict, total=False):
    text: str
    values: dict[str, object]
    data: dict[str, object]


class HealthReport(TypedDict):
    id: str
    name: str
    healthy: bool
    billing: str


async def get_container_health(
    auth: CloudAuthService | None = None,
    container_svc: CloudContainerService | None = None,
) -> ProviderResult:
    """Get ElizaCloud container health status."""
    if not auth or not auth.is_authenticated():
        return ProviderResult(text="")

    running = (
        [c for c in container_svc.get_tracked_containers() if c.status == "running"]
        if container_svc
        else []
    )
    if not running:
        return ProviderResult(
            text="No running containers.",
            values={"healthyContainers": 0},
        )

    reports: list[HealthReport] = [
        HealthReport(
            id=c.id,
            name=c.name,
            healthy=c.billing_status == "active",
            billing=c.billing_status,
        )
        for c in running
    ]

    healthy_count = len([r for r in reports if r["healthy"]])
    lines = [
        f"Health: {healthy_count}/{len(reports)} healthy",
        *[
            f"  - {r['name']}: {'OK' if r['healthy'] else 'UNHEALTHY'} ({r['billing']})"
            for r in reports
        ],
    ]

    return ProviderResult(
        text="\n".join(lines),
        values={
            "healthyContainers": healthy_count,
            "unhealthyContainers": len(reports) - healthy_count,
        },
        data={"reports": reports},  # type: ignore[typeddict-item]
    )


container_health_provider: dict[str, object] = {
    "name": "elizacloud_health",
    "description": "ElizaCloud container health",
    "dynamic": True,
    "position": 92,
    "private": True,
    "get": get_container_health,
}
